torchtitan/models/llama_multimodal/model.py

```diff
@@ -969,3 +996,520 @@ def forward(
+        xk = keys.transpose(1, 2)  # (bs, n_local_heads, seqlen, head_dim)
+        xv = values.transpose(1, 2)  # (bs, n_local_heads, seqlen, head_dim)
+
+        # we use casual mask for training
+        output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)
+        output = output.transpose(
+            1, 2
+        ).contiguous()  # (bs, seqlen, n_local_heads, head_dim)
+        output = output.view(bs, seqlen, -1)
+        return self.wo(output)
+
+
+class CrossAttention(nn.Module):
+    """
+    Multi-head cross attention module.
+
+    Args:
+        model_args (ModelArgs): Model configuration arguments.
+
+    Attributes:
+        num_kv_heads (int): Number of key and value heads.
+        num_heads (int): Number of query heads.
+        n_rep (int): Number of repetitions for local heads.
+        head_dim (int): Dimension size of each attention head.
+        wq (Linear): Linear transformation for queries.
+        wk (Linear): Linear transformation for keys.
+        wv (Linear): Linear transformation for values.
+        wo (Linear): Linear transformation for output.
+
+    """
+
+    def __init__(self, model_args: ModelArgs):
+        super().__init__()
+        self.num_heads = model_args.num_heads
+        self.num_kv_heads = (
+            model_args.num_heads
+            if model_args.num_kv_heads is None
+            else model_args.num_kv_heads
+        )
+        self.n_rep = self.num_heads // self.num_kv_heads
+        self.head_dim = model_args.dim // model_args.num_heads
+
+        self.wq = nn.Linear(
+            model_args.dim, model_args.num_heads * self.head_dim, bias=False
+        )
+        self.wk = nn.Linear(
+            model_args.dim, self.num_kv_heads * self.head_dim, bias=False
+        )
+        self.wv = nn.Linear(
+            model_args.dim, self.num_kv_heads * self.head_dim, bias=False
+        )
+        self.wo = nn.Linear(
+            model_args.num_heads * self.head_dim, model_args.dim, bias=False
+        )
+        self.q_norm = RMSNorm(dim=self.head_dim, eps=1e-05)
+        self.k_norm = RMSNorm(dim=self.head_dim, eps=1e-05)
+
+    def init_weights(self, init_std: float):
+        for linear in (self.wq, self.wk, self.wv):
+            nn.init.trunc_normal_(linear.weight, mean=0.0, std=0.02)
+        nn.init.trunc_normal_(self.wo.weight, mean=0.0, std=init_std)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        encoder_input: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+    ):
+        """
+        Forward pass of the attention module.
+
+        Args:
+            x (torch.Tensor): Input tensor.
+            encoder_input (torch.Tensor): Precomputed frequency tensor.
+
+        Returns:
+            torch.Tensor: Output tensor after attention.
+
+        """
+        bs, seqlen_x, _ = x.shape
+        seqlen_y = encoder_input.shape[1]
+        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
```

Comment:
should cross attention share the same input for all projections?

Reply:
oppps, my bad..